### DISCUSSION TITLE
December 2020 Update and Roadmap

### DISCUSSION THREAD

**Discussion by jstoecker** [microsoft/tensorflow-directml#131] [General]
As we wrap up 2020, our team wanted to give a brief update on the state of TensorFlow-DirectML and the direction we're headed for the new year.

We just released [TensorFlow-DirectML 1.15.4.dev201216](https://github.com/microsoft/tensorflow-directml/releases/tag/v1.15.4.dev201216), which focuses on addressing stability and low-hanging performance and memory issues. This update paves the way for more functionality and optimizations that should be coming early in 2021. Until then, test out the new packages (`pip install tensorflow-directml`) and let us know if you run into any issues. And, if you'd like try an application that uses TensorFlow with DirectML check out the exciting [new version of Lobe](https://lobe.ai/blog/four-new-features-in-lobe/), which now has GPU acceleration in Windows!

We have a lot more planned for 2021, and you can find a [roadmap](https://github.com/microsoft/tensorflow-directml/wiki/Roadmap) for what to expect in the coming months. We're also trying out GitHub's new _discussions_ feature, so feel free to ask questions here (continue to use _issues_ for bugs) and let us know what you think!

**Comment by Agrover112** [comment]
Will there be future support for TF2.x?

**Reply by jstoecker** [reply]
That's one of the things we're going to look at as soon we can finish TF 1.15. See more comments here: https://github.com/microsoft/tensorflow-directml/wiki/Roadmap#future-and-tensorflow-2

**Comment by ianier** [comment]
Any plans to support UWP? I've successfully compiled TF2.x for UWP targeting x64 (see https://github.com/microsoft/vcpkg/issues/14252#issuecomment-720606994), but unfortunately an ARM64 port isn't as easy. 
In the roadmap it's mentioned that ARM64 is one of your future targets, but it doesn't say anything about UWP.

**Reply by jstoecker** [reply]
What specific scenario did you have in mind?

We haven't scoped out the work required for UWP since our focus is accelerating training, which isn't commonly done in a UWP app. Windows already supports hardware-accelerated inference in UWP (using DirectML!) with other frameworks, and DirectX/DirectML works just fine in UWP apps, so the question is mostly about TF. We still want to upstream into mainline TF at some point and in some form (not an option for TF 1.x, but 2.x potentially), so it seems like the better place to do this work would be in the main TF repo. It really all depends on the demand/need for UWP support for TF.

**Reply by ianier** [reply]
Indeed, the scenario I have in mind is inference of existing TF models in a UWP app. There are a couple of choices:

**1-Convert the TF model to ONNX so it can run on WinML**: no go for complex models (https://github.com/onnx/tensorflow-onnx/issues/1108)
**2-UWP support in the TF main repo**: not enough interest (https://github.com/tensorflow/tensorflow/issues/44463)
**3-Patch TF to get it to compile for UWP**: this is what I did and it works, but only on x64 and it's currently a hack (https://github.com/microsoft/vcpkg/pull/14394)

I can live with option 3, but it's a pity that Windows on ARM64 isn't even supported by TF (https://github.com/tensorflow/tensorflow/issues/45151), so I was happy to read that ARM64 support is in your long term roadmap, and I'm curious about how you plan to achieve this.

In comparison, the day Apple released the M1 Mac mini, the TF community was already working on porting TF to run on it :-).

**Reply by jstoecker** [reply]
It looks like just supporting UWP wouldn't even address all your needs since your model also needs complex numbers (not yet available in DirectML or even ONNX) and ARM64. These features may come, but it's going to take some time.

ARM build support in TF will be interesting for us as well.  :)

**Reply by ianier** [reply]
My assumption was that the DirectML fork for TF 2.x would implement complex math, or at least emulate it. Also, hoping for UWP support wouldn't be unreasonable in a Microsoft-sponsored fork (though even a functional ARM64 build with VC++ would get me halfway there).
I'm an enthusiast with limited knowledge of the TF build system, but I do have access to ARM64 hardware and an [actual UWP app](https://www.microsoft.com/en-us/p/chronotron/9wzdncrfjkx2?activetab=pivot%3Aoverviewtab) as use case, so I'm willing to contribute the best I can.
